components/core/src/clp/CurlDownloadHandler.cpp

```diff
@@ -48,13 +58,47 @@ CurlDownloadHandler::CurlDownloadHandler(
     m_easy_handle.set_option(CURLOPT_TIMEOUT, static_cast<long>(overall_timeout.count()));
 
     // Set up http headers
+    constexpr std::string_view cRangeHeaderName{"range"};
+    constexpr std::string_view cCacheControlHeaderName{"cache-control"};
+    constexpr std::string_view cPragmaHeaderName{"pragma"};
+    std::unordered_set<std::string_view> reserved_headers;
+    reserved_headers.insert(cRangeHeaderName);
+    reserved_headers.insert(cCacheControlHeaderName);
+    reserved_headers.insert(cPragmaHeaderName);
     if (0 != offset) {
-        std::string const range{"Range: bytes=" + std::to_string(offset) + "-"};
-        m_http_headers.append(range);
+        m_http_headers.append(fmt::format("{}: bytes={}-", cRangeHeaderName, offset));
     }
     if (disable_caching) {
-        m_http_headers.append("Cache-Control: no-cache");
-        m_http_headers.append("Pragma: no-cache");
+        m_http_headers.append(fmt::format("{}: no-cache", cCacheControlHeaderName));
+        m_http_headers.append(fmt::format("{}: no-cache", cPragmaHeaderName));
+    }
+    if (http_header_kv_pairs.has_value()) {
+        for (auto const& [key, value] : http_header_kv_pairs.value()) {
+            // Convert to lowercase for case-insensitive comparison
```

Comment:
```suggestion
            // HTTP header field-name (key) is case-insensitive:
            // https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2
            // Therefore, we convert keys to the lowercase for comparison with the reserved keys.
```
Ideally, we don't validate whether user-input headers has any duplicated keys. But IMO we don't have to check this for two reasons:
1. The behavior on duplicated key handling is implementation-dependent on the server end
2. We only need to check reserved keys to ensure the APIs we provide have defined behaviors

Any comments? @kirkrodrigues

Reply:
Which keys could be duplicated? The user can't pass in duplicated keys since it's a map, right? Do you mean duplicates between what the user provides and what we provide?

Reply:
Ideally they can, because the actual field name in HTTP headers is case-insensitive. So they could pass in sth like `Key`, `key`, and `KEY` within the unordered map.